cluster/juju/layers/kubernetes-master/tactics/update_addons.py

```diff
@@ -54,32 +60,40 @@ def add_addon(source, dest):
     with open(dest, "w") as f:
         f.write(content)
 
+
 def update_addons(dest):
     """ Update addons. This will clean the addons folder and add new manifests
     from upstream. """
     clean_addon_dir(dest)
     with kubernetes_repo() as repo:
-        add_addon(repo + "/cluster/addons/dashboard/dashboard-controller.yaml", dest)
-        add_addon(repo + "/cluster/addons/dashboard/dashboard-service.yaml", dest)
-        add_addon(repo + "/cluster/addons/dns/kubedns-controller.yaml.in", dest + "/kubedns-controller.yaml")
-        add_addon(repo + "/cluster/addons/dns/kubedns-svc.yaml.in", dest + "/kubedns-svc.yaml")
-        add_addon(repo + "/cluster/addons/cluster-monitoring/influxdb/grafana-service.yaml", dest)
-        add_addon(repo + "/cluster/addons/cluster-monitoring/influxdb/heapster-controller.yaml", dest)
-        add_addon(repo + "/cluster/addons/cluster-monitoring/influxdb/heapster-service.yaml", dest)
-        add_addon(repo + "/cluster/addons/cluster-monitoring/influxdb/influxdb-grafana-controller.yaml", dest)
-        add_addon(repo + "/cluster/addons/cluster-monitoring/influxdb/influxdb-service.yaml", dest)
+        add_addon(repo + "/cluster/addons/dashboard/dashboard-controller.yaml",
+                  dest)
+        add_addon(repo + "/cluster/addons/dashboard/dashboard-service.yaml",
+                  dest)
+        add_addon(repo + "/cluster/addons/dns/kubedns-controller.yaml.in",
+                  dest + "/kubedns-controller.yaml")
+        add_addon(repo + "/cluster/addons/dns/kubedns-svc.yaml.in",
+                  dest + "/kubedns-svc.yaml")
+        influxdb = "/cluster/addons/cluster-monitoring/influxdb"
+        add_addon(repo + influxdb + "/grafana-service.yaml", dest)
+        add_addon(repo + influxdb + "/heapster-controller.yaml", dest)
+        add_addon(repo + influxdb + "/heapster-service.yaml", dest)
+        add_addon(repo + influxdb + "/influxdb-grafana-controller.yaml", dest)
+        add_addon(repo + influxdb + "/influxdb-service.yaml", dest)
 
 # Entry points
 
+
 class UpdateAddonsTactic(Tactic):
     """ This tactic is used by charm-tools to dynamically populate the
     template/addons folder at `charm build` time. """
 
     @classmethod
-    def trigger(cls, relpath):
+    def trigger(cls, entity, target, layer, next_config):
         """ Determines which files the tactic should apply to. We only want
         this tactic to trigger once, so let's use the templates/ folder
         """
+        relpath = entity.relpath(layer.directory) if layer else entity
```

Comment:
If the goal is to have this work with the older 2.1 branch of charm-tools, it doesn't:

```
$ charm build --no-local-layers
build: Composing into /home/gkk/work/juju-repo/charms
build: Destination charm directory: /home/gkk/work/juju-repo/charms/builds/kubernetes-master
build: Processing layer: layer:basic
build: Processing layer: layer:tls-client
build: Processing layer: layer:debug
build: Processing layer: kubernetes-master
Traceback (most recent call last):
  File "/home/gkk/.local/bin/charm-build", line 11, in <module>
    sys.exit(main())
  File "/usr/lib/python2.7/dist-packages/charmtools/build/builder.py", line 676, in main
    build()
  File "/usr/lib/python2.7/dist-packages/charmtools/build/builder.py", line 519, in __call__
    self.generate()
  File "/usr/lib/python2.7/dist-packages/charmtools/build/builder.py", line 470, in generate
    self.formulate_plan(layers)
  File "/usr/lib/python2.7/dist-packages/charmtools/build/builder.py", line 411, in formulate_plan
    self.plan = self.plan_layers(layers, output_files)
  File "/usr/lib/python2.7/dist-packages/charmtools/build/builder.py", line 327, in plan_layers
    output_files=output_files))
  File "/usr/lib/python2.7/dist-packages/charmtools/build/builder.py", line 323, in <genexpr>
    list(e for e in utils.walk(layer.directory,
  File "/usr/lib/python2.7/dist-packages/charmtools/utils.py", line 345, in walk
    yield (entry, fn(entry, **kwargs))
  File "/usr/lib/python2.7/dist-packages/charmtools/build/builder.py", line 296, in build_tactics
    current = current.config.tactic(entry, current, self.target, config)
  File "/usr/lib/python2.7/dist-packages/charmtools/build/config.py", line 111, in tactic
    if tactic.trigger(entity.relpath(bd)):
TypeError: trigger() takes exactly 5 arguments (2 given)

$ charm version
charm 2.2.0-0ubuntu1~ubuntu16.04.1~ppa2
charm-tools 2.1.9
```

I think if you add default params, it should work. Something like this:

```python
def trigger(cls, entity, target=None, layer=None, next_config=None):
```

Also, since I don't think it's obvious just from reading the code, can you add a comment to explain that this is for backward compatibility with a different signature?